### PR TITLE
Automate release process + misc chores

### PR DESCRIPTION
**Pull request by mikejpeters**
Replaces the `publish` action with a `release` action using [semantic-release](https://github.com/semantic-release/semantic-release). The `semantic-release` tool will generate a release and publish to NPM out-of-the-box, no configuration necessary.

The action is set up so it must be [run manually](https://docs.github.com/en/actions/managing-workflow-runs/manually-running-a-workflow). 

### NOTES 

* semantic-release takes care of updating the `package.json` version before publishing to npm, so the version on source control is no longer used. [FAQ](https://github.com/semantic-release/semantic-release/blob/master/docs/support/FAQ.md)
* the github action references GITHUB_TOKEN env var - but according to these [docs](https://github.com/semantic-release/semantic-release/blob/master/docs/recipes/ci-configurations/github-actions.md) this variable is automatically populated

Also includes changes to:

* ignore vscode project files
* rename integration test job from "build" to "test"
* use commitizen to help with commit messages
  (and document in CONTRIBUTING.md)
* use commitlint to lint commit messages
* move tooling config to separate files to avoid cluttering package.json
  and improve commit diffs

### CODE REVIEW

File: .czrc
Comment (fernando-mc) [inline .czrc:2]: Cool, haven't used this before. Should save some work with the release notes.

File: .github/workflows/release.yml
Comment (fernando-mc) [inline .github/workflows/release.yml:6]: I have no issues with running this manually vs. using the release workflow but I am curious why to go this route vs. the release trigger:
https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#release

Is this so we can use the automatic semver without issue? Either way, looks good.

File: package.json
Comment (fernando-mc) [inline package.json:3]: What's the reason for 0.0.0 here? Will it automatically check the current NPM semver? Or does this need to be updated?

**Review by fernando-mc — COMMENTED** [review]
Just a few questions, if no changes are needed from them consider this an approval.

File: .github/workflows/release.yml
Comment (mikejpeters) [reply .github/workflows/release.yml:6]: Because it creates the release automatically, so this way you just click an action and it will automatically determine the correct version and notes to put in the release. I think that's better than manually creating a release where it's possible to put the wrong version etc

File: package.json
Comment (mikejpeters) [reply package.json:3]: It uses the  latest tag on GitHub + the commits since then to determine the new release version before publishing to npm. The 0.0.0 in package json is recommended so we know it's not used or updated

File: .github/workflows/release.yml
Comment (fernando-mc) [reply .github/workflows/release.yml:6]: Perfect, makes sense to me. As long as we can still create GitHub release notes should be fine. At this point we can probably cleanup the readme and remove some of the old release notes or move them to the GitHub releases section anyway.

File: package.json
Comment (fernando-mc) [reply package.json:3]: Sounds good, thanks!

**Comment by fernando-mc** [comment]
Give @Shereef a chance to peek it over so we're up to speed on new changes but LGTM thanks for addressing questions.

**Review by fernando-mc — APPROVED** [review]

File: .husky/commit-msg
Comment (Shereef) [inline .husky/commit-msg:4]: I am trying to find what does --no but all I can find it --no-install and I can't find --no can you please help me @mikejpeters

Comment (Shereef) [reply .husky/commit-msg:4]: @mikejpeters  I also did:

i executed `npm i` and `npm run prepare` 
I then proceeded to edit the readme to make silly commit messages like these:

```
commit a365c96db1d6199f3993f5c022274857b688d66a (HEAD -> misc-chores)
Author: Shereef Marzouk <shereef.marzouk@daysmart.com>
Date:   Sun Feb 27 13:22:05 2022 -0400

    Tuh4iuyuifkjsdjklswkj 3

commit 789e487e68d691cd647c83e6fae6ee1e47e446e9
Author: Shereef Marzouk <shereef.marzouk@daysmart.com>
Date:   Sun Feb 27 13:21:41 2022 -0400

    KHGDJHGSD: kjashkjhgaskjd

commit c43cd57b9d531113074a5fc583b93661137640ab
Author: Shereef Marzouk <shereef.marzouk@daysmart.com>
Date:   Sun Feb 27 13:21:21 2022 -0400

    : ascasdsad

commit 376a58dfe4a3b9630f86c6f1c459c01ffe365e8b (origin/misc-chores)
Author: Mike Peters <mikejpeters@gmail.com>
Date:   Sat Feb 26 16:05:10 2022 -0800

    chore: automate release process
    
    closes #146
```

it committed fine 
What am I doing wrong ?

I feel like at least 2 of these go against `@commitlint/config-conventional`

I haven't used semantic-release, commitlint or commitizen before but I am looking forward to learning them

Comment (mikejpeters) [reply .husky/commit-msg:4]: By default if you try to run a package with `npx` that isn't installed, it will prompt you to proceed:

```
$ npx commitlint
Need to install the following packages:
  commitlint
Ok to proceed? (y)
```

But in the `commit-msg` hook, prompts wouldn't work correctly, so it's better to suppress the prompt (which can be done with `--yes` or `--no`). Here we prefer `--no` since the package should be installed with the rest of the dev dependencies, and if it isn't, better to get an error right away.

From [npx](https://docs.npmjs.com/cli/v8/commands/npx) docs:

> If any requested packages are not present in the local project dependencies, then they are installed to a folder in the npm cache, which is added to the PATH environment variable in the executed process. A prompt is printed (which can be suppressed by providing either --yes or --no).

File: .github/workflows/release.yml
Comment (Shereef) [inline .github/workflows/release.yml:28]: I tried running it locally it showed the following
it failed dry run for som reason do you know why ?
Do I have to create a tag first ?

```
Shmac:serverless-finch shereef.marzouk$ semantic-release
[1:30:27 PM] [semantic-release] › ℹ  Running semantic-release version 19.0.2
[1:30:27 PM] [semantic-release] › ✔  Loaded plugin "verifyConditions" from "@semantic-release/npm"
[1:30:27 PM] [semantic-release] › ✔  Loaded plugin "verifyConditions" from "@semantic-release/github"
[1:30:27 PM] [semantic-release] › ✔  Loaded plugin "analyzeCommits" from "@semantic-release/commit-analyzer"
[1:30:27 PM] [semantic-release] › ✔  Loaded plugin "generateNotes" from "@semantic-release/release-notes-generator"
[1:30:27 PM] [semantic-release] › ✔  Loaded plugin "prepare" from "@semantic-release/npm"
[1:30:27 PM] [semantic-release] › ✔  Loaded plugin "publish" from "@semantic-release/npm"
[1:30:27 PM] [semantic-release] › ✔  Loaded plugin "publish" from "@semantic-release/github"
[1:30:27 PM] [semantic-release] › ✔  Loaded plugin "addChannel" from "@semantic-release/npm"
[1:30:27 PM] [semantic-release] › ✔  Loaded plugin "addChannel" from "@semantic-release/github"
[1:30:27 PM] [semantic-release] › ✔  Loaded plugin "success" from "@semantic-release/github"
[1:30:27 PM] [semantic-release] › ✔  Loaded plugin "fail" from "@semantic-release/github"
[1:30:27 PM] [semantic-release] › ⚠  This run was not triggered in a known CI environment, running in dry-run mode.
[1:30:29 PM] [semantic-release] › ✖  An error occurred while running semantic-release: Error: Command failed with exit code 1: git fetch --tags --update-head-ok https://github.com/fernando-mc/serverless-finch.git +refs/heads/master:refs/heads/master
From https://github.com/fernando-mc/serverless-finch
 ! [rejected]        v3.0.0     -> v3.0.0  (would clobber existing tag)
    at makeError (/usr/local/lib/node_modules/semantic-release/node_modules/execa/lib/error.js:60:11)
    at handlePromise (/usr/local/lib/node_modules/semantic-release/node_modules/execa/index.js:118:26)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at async fetch (/usr/local/lib/node_modules/semantic-release/lib/git.js:120:5)
    at async /usr/local/lib/node_modules/semantic-release/lib/branches/index.js:21:5
    at async pEachSeries (/usr/local/lib/node_modules/semantic-release/node_modules/p-each-series/index.js:8:23)
    at async module.exports (/usr/local/lib/node_modules/semantic-release/lib/branches/index.js:20:3)
    at async run (/usr/local/lib/node_modules/semantic-release/index.js:65:22)
    at async module.exports (/usr/local/lib/node_modules/semantic-release/index.js:268:22)
    at async module.exports (/usr/local/lib/node_modules/semantic-release/cli.js:55:5) {
  shortMessage: 'Command failed with exit code 1: git fetch --tags --update-head-ok https://github.com/fernando-mc/serverless-finch.git +refs/heads/master:refs/heads/master',
  command: 'git fetch --tags --update-head-ok https://github.com/fernando-mc/serverless-finch.git +refs/heads/master:refs/heads/master',
  escapedCommand: 'git fetch --tags --update-head-ok "https://github.com/fernando-mc/serverless-finch.git" "+refs/heads/master:refs/heads/master"',
  exitCode: 1,
  signal: undefined,
  signalDescription: undefined,
  stdout: '',
  stderr: 'From https://github.com/fernando-mc/serverless-finch\n' +
    ' ! [rejected]        v3.0.0     -> v3.0.0  (would clobber existing tag)',
  failed: true,
  timedOut: false,
  isCanceled: false,
  killed: false
}
Error: Command failed with exit code 1: git fetch --tags --update-head-ok https://github.com/fernando-mc/serverless-finch.git +refs/heads/master:refs/heads/master
From https://github.com/fernando-mc/serverless-finch
 ! [rejected]        v3.0.0     -> v3.0.0  (would clobber existing tag)
    at makeError (/usr/local/lib/node_modules/semantic-release/node_modules/execa/lib/error.js:60:11)
    at handlePromise (/usr/local/lib/node_modules/semantic-release/node_modules/execa/index.js:118:26)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at async fetch (/usr/local/lib/node_modules/semantic-release/lib/git.js:120:5)
    at async /usr/local/lib/node_modules/semantic-release/lib/branches/index.js:21:5
    at async pEachSeries (/usr/local/lib/node_modules/semantic-release/node_modules/p-each-series/index.js:8:23)
    at async module.exports (/usr/local/lib/node_modules/semantic-release/lib/branches/index.js:20:3)
    at async run (/usr/local/lib/node_modules/semantic-release/index.js:65:22)
    at async module.exports (/usr/local/lib/node_modules/semantic-release/index.js:268:22)
    at async module.exports (/usr/local/lib/node_modules/semantic-release/cli.js:55:5) {
  shortMessage: 'Command failed with exit code 1: git fetch --tags --update-head-ok https://github.com/fernando-mc/serverless-finch.git +refs/heads/master:refs/heads/master',
  command: 'git fetch --tags --update-head-ok https://github.com/fernando-mc/serverless-finch.git +refs/heads/master:refs/heads/master',
  escapedCommand: 'git fetch --tags --update-head-ok "https://github.com/fernando-mc/serverless-finch.git" "+refs/heads/master:refs/heads/master"',
  exitCode: 1,
  signal: undefined,
  signalDescription: undefined,
  stdout: '',
  stderr: 'From https://github.com/fernando-mc/serverless-finch\n' +
    ' ! [rejected]        v3.0.0     -> v3.0.0  (would clobber existing tag)',
  failed: true,
  timedOut: false,
  isCanceled: false,
  killed: false
```

File: .husky/commit-msg
Comment (mikejpeters) [reply .husky/commit-msg:4]: Hm I'm not sure, can you please share the output from `git commit`?

File: .github/workflows/release.yml
Comment (Shereef) [reply .github/workflows/release.yml:28]: I manually created v3.0.1 locally and ran it again it gave the same exact output

**Comment by Shereef** [comment]
My comments above are not requesting changes or declining the PR. Just seeking clarification and making sure everything works as expected.

File: .github/workflows/release.yml
Comment (mikejpeters) [reply .github/workflows/release.yml:28]: I don't know for sure, but I think this is only happening if your local branch/tags are not in sync with the remote for some reason (i.e. a force push). I don't think this should happen when it runs on CI

File: .husky/commit-msg
Comment (Shereef) [reply .husky/commit-msg:4]:
```
[2022-02-27T17:39:19.659Z] > git add -A -- . [70ms]
[2022-02-27T17:39:19.706Z] > git -c user.useConfigOnly=true commit --quiet --allow-empty-message --file - [46ms]
[2022-02-27T17:39:19.706Z] hint: The '.husky/pre-commit' hook was ignored because it's not set as executable.
hint: You can disable this warning with `git config advice.ignoredHook false`.
hint: The '.husky/commit-msg' hook was ignored because it's not set as executable.
hint: You can disable this warning with `git config advice.ignoredHook false`.
[2022-02-27T17:39:19.751Z] > git symbolic-ref --short HEAD [19ms]
[2022-02-27T17:39:19.776Z] > git for-each-ref --format=%(refname)%00%(upstream:short)%00%(objectname)%00%(upstream:track) refs/heads/misc-chores refs/remotes/misc-chores [25ms]
[2022-02-27T17:39:19.802Z] > git remote --verbose [25ms]
[2022-02-27T17:39:19.901Z] > git for-each-ref --sort -committerdate --format %(refname) %(objectname) %(*objectname) [124ms]
[2022-02-27T17:39:19.922Z] > git config --get commit.template [20ms]
[2022-02-27T17:39:19.941Z] > git config --get commit.template [17ms]
[2022-02-27T17:39:20.941Z] > git ls-files --stage -- /Users/shereef.marzouk/Dev/serverless-finch/README.md [65ms]
[2022-02-27T17:39:20.983Z] > git cat-file -s 79b4ecde7f27fe785fcbd2d561dae0d7697c1fb1 [40ms]
```
this is my VSCode output in the git tab


I see 
hint: The '.husky/pre-commit' hook was ignored because it's not set as executable.
hint: You can disable this warning with `git config advice.ignoredHook false`.

File: .github/workflows/release.yml
Comment (Shereef) [reply .github/workflows/release.yml:28]: ok we will just have to test it after merging with the next release

**Comment by mikejpeters** [comment]
Shoot I really didn't mean to merge :( sorry I will fix the issue with permissions on `commit-msg` now

File: .husky/commit-msg
Comment (mikejpeters) [reply .husky/commit-msg:4]: I messed up and merged this :( But I've fixed the hook permissions on master. Can you pull the latest version and confirm that the hooks run for you?

**Comment by fernando-mc** [comment]
@mikejpeters no worries, it happens. That's why we have the manual release trigger instead of completely automated. Just try to get sign off next time. I'll just eventually migrate this to an Org so we can enforce approvals.

**Comment by fernando-mc** [comment]
Actually, looks like I can do this without Organization settings. Updating this now.

**Pull request merged**
